Validators

MarkGibbons opened the pull request:
I tried to add the documentation for the resource validator parameters.  This was my first try at a documentation update and I'm sorry I don't understand the structure as well as I should.  

I think it's pretty confusing not to be able to find the old validators in the custom resource documentation.  I can't tell if chef plans to drop them, but until then I would like to be able to find them.

Thanks,
  Mark

Commit:
Add the property validation methods

Properties have some validation methods that were not documented.  Added and
modified the documentation of the validation methods from pre 12.5.1
resource attributes.  Documentation for the missing methods was too hard to find unless
you knew it existed. It was also not obvious that the methods had been carried forward
with the new style custom resources.

Commit:
Change the location of the validator documentation file

Commit:
Make all the dsl_custom_resource_method_property tagged areas consistent

Tensibai commented on snapshots/includes_dsl_custom_resource/includes_dsl_custom_resource_method_property_attribute_validation_parameter.rst:
```diff
@@ -0,0 +1,113 @@
+
+.. code-block:: ruby
+
+   property :enabled, equal_to: [true, false, 'true', 'false'], default: true
```
Seems it is missing the Ruby type of the property before the validator part.

Tensibai commented on snapshots/includes_dsl_custom_resource/includes_dsl_custom_resource_method_property_attribute_validation_parameter.rst:
```diff
@@ -0,0 +1,113 @@
+.. The contents of this file should be modified in a way that preserves its ability to appear in multiple topics.
+
+
+A validation parameter is used to add zero (or more) validation parameters to an property.
```
Maybe `to a property` ? (unsure)

Tensibai reviewed (commented):
Few small points

Commit:
Changed the order of documentation sections for the property parameters.

Moved the validation parameters inline to a tagged section dsl_custom_resource_method_property_validation_parameter.
Fixed grammar.

MarkGibbons commented:
@Tensibai Thank you.  Should be fixed and tighter now.  Also got rid of the included file and created a new tagged section for the validator parameters.

Commit:
Remove errant character.

Tensibai commented on chef_master/source/custom_resources.rst:
```diff
@@ -857,7 +980,7 @@ For example, the following properties define ``username`` and ``password`` prope
 
 Block Arguments
-+++++++++++++++++++++++++++++++++++++++++++++++++++++
+-----------------------------------------------------
```
Why a different header size here ? This would break the overall presentation, or did I miss something ?

MarkGibbons replied:
It seemed to me that Block Arguments described something about properties that was useful in load_current_value processing and so should have the same header as property_is_set? and reset_property.  Grouping Block Arguments under the property section was confusing.  I'm willing to change this.

Tensibai replied:
Fair enough :)

Tensibai commented on chef_master/source/custom_resources.rst:
```diff
@@ -817,6 +818,128 @@ For example, the following properties define ``username`` and ``password`` prope
+
+.. code-block:: ruby
+
+   property :enabled, equal_to: [true, false, 'true', 'false'], default: true
```
Still missing a ruby type here, shouldn't it be `property :enabled, [TrueClass, FalseClass, String], equal_to: [true, false, 'true', 'false'], default: true` According to the above data (or if the type can be omitted, this worth a note even if I think it's a bad idea to not specify things)

MarkGibbons replied:
Updated the property description to note that ruby_type is optional.

Tensibai commented on snapshots/includes_dsl_custom_resource/includes_dsl_custom_resource_method_property.rst:
```diff
@@ -6,13 +6,14 @@ Use the ``property`` method to define properties for the custom resource. The sy
-* ``default: 'value'`` is the default value loaded into the resource
+* ``ruby_type`` is the |ruby| type, such as ``String``, ``Integer``, ``TrueClass``, or ``FalseClass``
+* ``default: 'value'`` is the optional default value loaded into the resource
+* ``validators: 'values'`` are optional validation parameters and values
```
Worth keeping the `parameter`  name here for coherence with the rest of the doc.

MarkGibbons replied:
Wrapped this section in .. tag dsl_custom_resource_method_property to keep it coherent.

Tensibai reviewed (commented):
The two first comments are valid for each file. Maybe worth using parameter_name instead to be clear it is not a syntax like `property :test_name, String, parameters: equal_to: [true,false]`

Commit:
Note that ruby_type is optional.

MarkGibbons commented:
@Tensibai Thank you again for another review.

Commit:
Removed start and end dtags

All docs under `/snapshots` are still built using multiple include files and do not reflect the recent change to the dtag syntax used in the `chef_master` directory. Therefore, we don't need to add in this tag. After we flatten these snapshot dirs, then new tags will be created for any shared content as part of the flattening process.

davidwrede commented:
![gif-keyboard-17449552733145657989](https://cloud.githubusercontent.com/assets/1634752/19823554/1051d142-9d1e-11e6-8bd8-a3046db6692b.gif)

Thanks for getting this added back in! 

I will check with our engineers to see if the content for param validators should be deprecated. If so, I will add a note in the appropriate sections.
